File: epfl-hc/hc2012/graders/CryptoLib1.java
import java.math.BigInteger;
import java.util.Scanner;

public class CryptoLib1 {

  private static Crypto1 crypto1;
  private static Scanner scanner;

  public static void main(String args[]) {
    crypto1 = new Crypto1();
    scanner = new Scanner(System.in);

    testDecrypt();

    System.out.println();
    System.out.println("OK - 54ec725043cfc34e86f1e2f62b84d280");
  }

  private static void unitTest() {
    BigInteger p = BigInteger.valueOf(1000003);
    BigInteger g = BigInteger.valueOf(2);
    BigInteger x = BigInteger.valueOf(4242);
    BigInteger r = BigInteger.valueOf(123456);
    BigInteger m = BigInteger.valueOf(654321);
    BigInteger y = g.modPow(x, p);

    PublicKey pk = new PublicKey();
    pk.p = p;
    pk.g = g;
    pk.y = y;

    Ciphertext ct = crypto1.encrypt(m, pk, r);
    if(! ct.E.equals(BigInteger.valueOf(509777))) {
      System.err.println("Reciprocal test failed (ct.E).");
      System.exit(0);
    }
    if(! ct.F.equals(BigInteger.valueOf(709498))) {
      System.err.println("Reciprocal test failed (ct.F).");
      System.exit(0);
    }
    System.out.println(ct.E + " " + ct.F);
    BigInteger mp = crypto1.decrypt(ct, pk, x);
    if (! mp.equals(m)) {
      System.err.println("Reciprocal test failed (m').");
      System.exit(0);
    } else {
      System.out.println("Reciprocal test ok.");
    }
  }

  private static void testDecrypt()
  {
    /*
      Input format is:
      - Number of testcases
      - For each testcase:   E  F  p  g  y  x  m  m2
    */
    int ntc = scanner.nextInt();
    for(int i=0;i<ntc;++i) {
      Ciphertext ct = new Ciphertext();
      ct.E = new BigInteger(scanner.next());
      ct.F = new BigInteger(scanner.next());
      PublicKey pk = new PublicKey();
      pk.p = new BigInteger(scanner.next());
      pk.g = new BigInteger(scanner.next());
      pk.y = new BigInteger(scanner.next());
      BigInteger x = new BigInteger(scanner.next());
      if(! pk.g.modPow(x, pk.p).equals(pk.y)) {
        System.err.println("ERROR: g^x != y  (mod p)");
        System.exit(1);
      }

      BigInteger m = crypto1.decrypt(ct, pk, x);
      BigInteger expectedM = new BigInteger(scanner.next());
      scanner.next(); // Remove m2 which we don't need for this test.
      if(! m.equals(expectedM)) {
        System.out.println();
        System.out.println("Testcase " + i + "  Wrong answer: expected " + expectedM + " got:");
        System.out.println(m);
        System.exit(0);
      } else {
        System.out.print(".");
      }
    }
  }

}

class Ciphertext {
  public BigInteger E;
  public BigInteger F;
}

class PublicKey {
  public BigInteger p;
  public BigInteger g;
  public BigInteger y;
}

File: epfl-hc/hc2012/graders/cryptoLib1.c
#include <stdio.h>
#include <stdlib.h>
#include <gmp.h>

#include "cryptoLib1.h"

#if 0
static void unitTest() {
    mpz_t x, r;
    PublicKey pk;
    Plaintext orig;

    mpz_init_set_si(pk.p, 1000003);
    mpz_init_set_si(pk.g, 2);
    mpz_init_set_si(x, 4242);
    mpz_init_set_si(r, 123456);
    mpz_init_set_si(orig.m, 654321);
    mpz_init(pk.y);

    mpz_powm(pk.y, pk.g, x, pk.p);

    Ciphertext ct = encrypt(orig, pk, r);
    if(mpz_get_si(ct.E) != 509777) {
      printf("Encrypt failed (wrong ct.E).\n");
      exit(0);
    }
    if(mpz_get_si(ct.F) != 709498) {
      printf("Encrypt failed (wrong ct.F).\n");
      exit(0);
    }
    Plaintext res = decrypt(ct, pk, x);
    if (mpz_cmp(res.m, orig.m)!=0) {
      printf("Decrypt failed (m' != m)\n");
      gmp_printf("Expected: %Zd\n", orig.m);
      gmp_printf("Got:      %Zd\n", res.m);
      exit(0);
    } else {
      printf("Unit test OK\n");
    }
}
#endif

static void testDecrypt() {
    /*
      Input format is:
      - Number of testcases
      - For each testcase:   E  F  p  g  y  x  m  m2
    */
    int ntc, i;
    if(1!=scanf("%d\n", &ntc)) {
      printf("Input failure.\n");
      exit(1);
    }
    for(i=0;i<ntc;++i) {
      Ciphertext ct;
      PublicKey pk;
      Plaintext got;
      mpz_t x;
      Plaintext m1, m2;
      mpz_init(ct.E);
      mpz_init(ct.F);
      mpz_init(pk.p);
      mpz_init(pk.g);
      mpz_init(pk.y);
      mpz_init(x);
      mpz_init(m1.m);
      mpz_init(m2.m);
      if(8!=gmp_scanf("%Zd %Zd %Zd %Zd %Zd %Zd %Zd %Zd\n", ct.E, ct.F, pk.p, pk.g, pk.y, x, m1.m, m2.m)) {
        printf("Input failure %d.\n", i);
        exit(1);
      }

      got = decrypt(ct, pk, x);
      if(mpz_cmp(got.m, m1.m)!=0) {
        gmp_printf("\nTestcase %d  Wrong answer: expected %Zd got:\n%Zd\n", i, m1.m, got.m);
        exit(0);
      } else {
        printf(".");
      }
      mpz_clear(got.m);
      mpz_clear(ct.E);
      mpz_clear(ct.F);
      mpz_clear(pk.p);
      mpz_clear(pk.g);
      mpz_clear(pk.y);
      mpz_clear(x);
      mpz_clear(m1.m);
      mpz_clear(m2.m);
    }
}

int main(int argc, char** argv) {
  testDecrypt();

  printf("\n");
  printf("OK - 54ec725043cfc34e86f1e2f62b84d280\n");
  return 0;
}
